src/python/ksc/rewrites.py

```diff
@@ -493,16 +517,18 @@ def visit_var(self, v: Var, var_names_to_exprs: VariableSubstitution):
 
     def visit_let(self, l: Let, var_names_to_exprs: VariableSubstitution) -> Let:
         assert isinstance(l.vars, Var), "use untuple_lets first"
+        rhs = self.visit(l.rhs, var_names_to_exprs)
+        # Update substitution for body
         target_var, var_names_to_exprs = _maybe_add_binder_to_subst(
             l.vars, var_names_to_exprs, [l.body]
         )
-        # Substitute bound var with target_var in children. It's fine to apply this substitution outside
-        # where the bound var is bound, as the replacement shouldn't contain "(let x ...) x" (with x free).
+        if target_var.type_.contains_any():
+            # Unfortunately we must set the type of target_var here, before recursing into the body,
+            # as said type depends on the binding here and can't be figured out from the body.
+            target_var.type_ = rhs.type_
```

Comment:
What's a rule that needs a symtab?  I assume we might mean
```clojure
(rule "sqrt_of_mul" (x : Float)
        (sqrt (mul x x)) 
        (abs x))
```
Which refers to `sqrt`, `mul`, `add`.  Then yes, absolutely exprs applying this rule should be using the same symtab (where I don't distinguish the symtab containing top-level definitions, and any local symtab).  Or did you mean something else?

Reply:
Yes, exactly. So I think this would mean an assert/check when we applied a rule, and then a lot of plumbing to (a) make tests pass that assert, (b) any client with its own symtab (e.g. from a user's program) to construct all the rule objects using that symtab.

This can be done; files like rewrites.py would probably no longer contain any RuleMatcher objects, only RuleMatcher subclasses and factories thereof (taking a symtab).